onnxruntime/core/providers/rocm/math/matmul_impl.cc

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+// Modifications: Remove cudaDeviceProp in LaunchFastGeluKernel.
+// Copyright (c) Advanced Micro Devices, Inc. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/rocm/math/matmul_impl.h"
+#include "core/providers/rocm/rocm_allocator.h"
+#include "core/providers/rocm/rocm_kernel.h"
+
+using namespace onnxruntime::rocm;
+
+namespace onnxruntime {
+namespace rocm {
+
+// StridedBatchedGemm can be used for the following GEMM computation
+// C[pnm] = A[pnk]*B[km] or C[pnm] = A[pnk]*B[pkm]
+static bool CanUseStridedBatchedGemm(const TensorShape& left_shape,
+                                     const TensorShape& right_shape,
+                                     bool transa, bool transb,
+                                     bool trans_batch_a, bool trans_batch_b,
+                                     int64_t& stride_A, int64_t& stride_B,
+                                     int64_t& stride_C, int64_t& batch_count) {
+  size_t left_num_dims = left_shape.NumDimensions();
+  size_t right_num_dims = right_shape.NumDimensions();
+
+  if (!(left_num_dims >= 3 && right_num_dims >= 2)) {
+    return false;
+  }
+
+  size_t left_leading_axis = trans_batch_a ? 0 : left_num_dims - 2;
+  size_t right_leading_axis = trans_batch_b ? 0 : right_num_dims - 2;
+  int64_t left_p = left_shape.SizeToDimension(left_num_dims - 2);
+  if (trans_batch_a) {
+    left_p = left_p * left_shape[left_num_dims - 2] / left_shape[0];
+  }
+  int64_t left_k =
+      transa ? left_shape[left_leading_axis] : left_shape[left_num_dims - 1];
+
+  if (right_num_dims >= 3) {
+    int64_t right_p = right_shape.SizeToDimension(right_num_dims - 2);
+    if (trans_batch_b) {
+      right_p = right_p * right_shape[right_num_dims - 2] / right_shape[0];
+    }
+    if (left_p != right_p) {
+      return false;
+    }
+  }
+
+  int64_t right_k = transb ? right_shape[right_num_dims - 1]
+                           : right_shape[right_leading_axis];
+  if (left_k != right_k) {
+    return false;
+  }
+
+  int64_t n =
+      transa ? left_shape[left_num_dims - 1] : left_shape[left_leading_axis];
+  int64_t m = transb ? right_shape[right_leading_axis]
+                     : right_shape[right_num_dims - 1];
+  stride_A = n * left_k / (trans_batch_a ? left_shape[0] : 1);
+  stride_B = right_num_dims == 2 ? 0 : right_k * m / (trans_batch_b ? right_shape[0] : 1);
+  stride_C = n * m;
+  batch_count = left_p;
+  return true;
+}
+
+template <typename T>
+Status MatMulImpl(const RocmKernel* op, MatMulComputeHelper& helper,
```

Comment:
Can helper be passed in as const reference? Not sure if we use pass by reference in the ORT codebase. Can you investigate a bit?

Reply:
helper is changed in this function, so we cannot use const reference. I saw some non-const reference in our codebase, so it might be a accecptable use.